x-pack/plugins/monitoring/server/alerts/thread_pool_rejections_alert_base.ts

```diff
@@ -0,0 +1,336 @@
+      values: { type },
+    });
+  }
+
+  protected static createActionVariables(type: string) {
+    return [
+      {
+        name: 'count',
+        description: i18n.translate(
+          'xpack.monitoring.alerts.threadPoolRejections.actionVariables.count',
+          {
+            defaultMessage: 'The number of nodes reporting high thread pool {type} rejections.',
+            values: { type },
+          }
+        ),
+      },
+      ...Object.values(AlertingDefaults.ALERT_TYPE.context),
+    ];
+  }
+
+  protected defaultParams: ThreadPoolRejectionsAlertParams = {
+    threshold: 300,
+    duration: '5m',
+  };
+
+  constructor(
+    rawAlert: Alert | undefined = undefined,
+    public readonly type: string,
+    public readonly threadPoolType: string,
+    public readonly label: string,
+    public readonly actionVariables: ActionVariables
+  ) {
+    super(rawAlert);
+  }
+
+  protected async fetchData(
+    params: ThreadPoolRejectionsAlertParams,
+    callCluster: any,
+    clusters: AlertCluster[],
+    uiSettings: IUiSettingsClient,
+    availableCcs: string[]
+  ): Promise<AlertData[]> {
+    let esIndexPattern = appendMetricbeatIndex(this.config, INDEX_PATTERN_ELASTICSEARCH);
+    if (availableCcs) {
+      esIndexPattern = getCcsIndexPattern(esIndexPattern, availableCcs);
+    }
+
+    const { threshold, duration } = params;
+
+    const stats = await fetchThreadPoolRejectionStats(
+      callCluster,
+      clusters,
+      esIndexPattern,
+      this.config.ui.max_bucket_size,
+      this.threadPoolType,
+      duration
+    );
+
+    return stats.map((stat) => {
+      const { clusterUuid, nodeId, rejectionCount, ccs } = stat;
+
+      return {
+        instanceKey: `${clusterUuid}:${nodeId}`,
+        shouldFire: rejectionCount > threshold,
+        rejectionCount,
+        severity: AlertSeverity.Danger,
+        meta: stat,
+        clusterUuid,
+        ccs,
+      };
+    });
+  }
+
+  protected filterAlertInstance(alertInstance: RawAlertInstance, filters: CommonAlertFilter[]) {
+    const alertInstanceStates = alertInstance.state
+      ?.alertStates as AlertThreadPoolRejectionsState[];
+    const nodeUuid = filters?.find((filter) => filter.nodeUuid)?.nodeUuid;
+
+    if (!alertInstanceStates?.length || !nodeUuid) {
+      return true;
+    }
+
+    const nodeAlerts = alertInstanceStates.filter(({ nodeId }) => nodeId === nodeUuid);
+    return Boolean(nodeAlerts.length);
+  }
+
+  protected getUiMessage(
+    alertState: AlertThreadPoolRejectionsState,
+    rejectionCount: number
+  ): AlertMessage {
+    const { nodeName, nodeId, type } = alertState;
+    return {
+      text: i18n.translate('xpack.monitoring.alerts.threadPoolRejections.ui.firingMessage', {
+        defaultMessage: `Node #start_link{nodeName}#end_link is reporting {rejectionCount} {type} rejections at #absolute`,
+        values: {
+          nodeName,
+          type,
+          rejectionCount,
+        },
+      }),
+      nextSteps: [
```

Comment:
We show charts for read and indexing threads in the node -> advanced page so we should probably link to that here too.